File: 283.move-zeroes.java
/*
 * @lc app=leetcode id=283 lang=java
 *
 * [283] Move Zeroes
 */
class Solution {

    // Assume nums is sparse
    public void moveZeroes(int[] nums) {
        if (nums == null) return;
        int mainIdx = 0;
        for (int runningIdx = 0; runningIdx < nums.length; runningIdx++) {
            if (nums[runningIdx] != 0) {
                if (runningIdx != mainIdx) {
                    nums[mainIdx] = nums[runningIdx];
                    nums[runningIdx] = 0;
                }
                mainIdx++;
            }
        }
    }
    
}

File: 283.move-zeroes.go
/*
 * @lc app=leetcode id=283 lang=golang
 *
 * [283] Move Zeroes
 */

// @lc code=start
func moveZeroes(nums []int)  {
	if nums == nil || len(nums) == 0 {
		return	
	}
	boxIdx := 0
	for ballIdx := 0; ballIdx < len(nums); ballIdx++ {
		if nums[ballIdx] != 0 {
			if boxIdx < ballIdx {
				nums[boxIdx] = nums[ballIdx]
				nums[ballIdx] = 0
			}
			boxIdx++
		}
	}
}
// @lc code=end

